### PR TITLE
Resolve remaining deprecated notice and passing null parameter

### PR DESCRIPTION
**Pull request by longthanhtran**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #19139

### CODE REVIEW

File: framework/BaseYii.php [hunk 1]
```diff
@@ -130,7 +130,7 @@ public static function getVersion()
      */
     public static function getAlias($alias, $throwException = true)
     {
-        if (strncmp($alias, '@', 1) !== 0) {
+        if (strncmp(isset($alias) ? $alias : '', '@', 1) !== 0) {
```
Comment (samdark) [inline framework/BaseYii.php:133]:
```suggestion
        if (strncmp((string)$alias), '@', 1) !== 0) {
```

File: framework/base/ExitException.php [hunk 2]
```diff
@@ -33,6 +33,8 @@ class ExitException extends \Exception
     public function __construct($status = 0, $message = null, $code = 0, $previous = null)
     {
         $this->statusCode = $status;
+        $message  = isset($message) ? $message : '';
+        $previous = isset($previous) ? $previous : new \Exception();
         parent::__construct($message, $code, $previous);
```
Comment (samdark) [inline framework/base/ExitException.php:38]:
```suggestion
        parent::__construct((string)$message, $code, $previous);
```

Did `$previous` stopped to accept `null`?

File: framework/base/Security.php [hunk 3]
```diff
@@ -298,6 +298,10 @@ protected function decrypt($data, $passwordBased, $secret, $info)
     public function hkdf($algo, $inputKey, $salt = null, $info = null, $length = 0)
     {
         if (function_exists('hash_hkdf')) {
+            $algo     = isset($algo) ? $algo : '';
+            $inputKey = isset($inputKey) ? $inputKey : '';
+            $salt     = isset($salt) ? $salt : '';
+            $info     = isset($info) ? $info : '';
             $outputKey = hash_hkdf($algo, $inputKey, $length, $info, $salt);
```
Comment (samdark) [inline framework/base/Security.php:305]:
```suggestion
            $outputKey = hash_hkdf((string)$algo, (string)$inputKey, $length, (string)$info, (string)$salt);
```

File: framework/console/controllers/HelpController.php [hunk 4]
```diff
@@ -473,8 +473,8 @@ protected function getSubCommandHelp($controller, $actionID)
      */
     protected function formatOptionHelp($name, $required, $type, $defaultValue, $comment)
     {
-        $comment = trim($comment);
-        $type = trim($type);
+        $comment = trim(isset($comment) ? $comment : '');
+        $type = trim(isset($type) ? $type : '');
```
Comment (samdark) [inline framework/console/controllers/HelpController.php:477]:
```suggestion
        $comment = trim((string)$comment);
        $type = trim((string)$type);
```

File: framework/helpers/BaseConsole.php [hunk 5]
```diff
@@ -331,6 +331,7 @@ public static function xtermBgColor($colorCode)
      */
     public static function stripAnsiFormat($string)
     {
+        $string = isset($string) ? $string : '';
         return preg_replace(self::ansiCodesPattern(), '', $string);
```
Comment (samdark) [inline framework/helpers/BaseConsole.php:335]:
```suggestion
        return preg_replace(self::ansiCodesPattern(), '', (string)$string);
```

File: framework/helpers/BaseConsole.php [hunk 6]
```diff
@@ -370,6 +371,7 @@ public static function ansiColorizedSubstr($string, $start, $length)
             return '';
         }
 
+        $string = isset($string) ? $string : '';
```
Comment (samdark) [inline framework/helpers/BaseConsole.php:374]:
```suggestion
        $string = (string)$string;
```

File: framework/helpers/BaseFileHelper.php [hunk 7]
```diff
@@ -900,7 +901,7 @@ protected static function normalizeOptions(array $options)
      */
     public static function changeOwnership($path, $ownership, $mode = null)
     {
-        if (!file_exists($path)) {
+        if (!file_exists(isset($path) ? $path : '')) {
```
Comment (samdark) [inline framework/helpers/BaseFileHelper.php:904]:
```suggestion
        if (!file_exists((string)$path)) {
```

File: framework/helpers/BaseHtml.php [hunk 8]
```diff
@@ -115,6 +115,7 @@ class BaseHtml
      */
     public static function encode($content, $doubleEncode = true)
     {
+        $content = isset($content) ? $content : '';
         return htmlspecialchars($content, ENT_QUOTES | ENT_SUBSTITUTE, Yii::$app ? Yii::$app->charset : 'UTF-8', $doubleEncode);
```
Comment (samdark) [inline framework/helpers/BaseHtml.php:119]:
```suggestion
        return htmlspecialchars((string)$content, ENT_QUOTES | ENT_SUBSTITUTE, Yii::$app ? Yii::$app->charset : 'UTF-8', $doubleEncode);
```

File: framework/helpers/BaseStringHelper.php [hunk 9]
```diff
@@ -29,7 +29,7 @@ class BaseStringHelper
      */
     public static function byteLength($string)
     {
-        return mb_strlen($string, '8bit');
+        return mb_strlen(isset($string) ? $string : '', '8bit');
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php:32]:
```suggestion
        return mb_strlen((string)$string, '8bit');
```

File: framework/i18n/GettextMoFile.php [hunk 10]
```diff
@@ -106,7 +106,7 @@ public function load($filePath, $context)
         $messages = [];
         for ($i = 0; $i < $count; ++$i) {
             $id = $this->readString($fileHandle, $sourceLengths[$i], $sourceOffsets[$i]);
-            $separatorPosition = strpos($id, chr(4));
+            $separatorPosition = strpos(isset($id) ? $id : '', chr(4));
```
Comment (samdark) [inline framework/i18n/GettextMoFile.php:109]:
```suggestion
            $separatorPosition = strpos((string)$id, chr(4));
```

**Review by samdark — CHANGES_REQUESTED** [review]
String checks are better to be done with type-casting.

File: framework/base/ExitException.php (hunk 2 above)
Comment (longthanhtran) [reply framework/base/ExitException.php:38]: in this case, the $previous requires an instance of Throwable so I use the `new \Exception()`

Comment (samdark) [reply framework/base/ExitException.php:38]: Then it's better to do:

```php
if ($previous === null) {
    parent::__construct((string)$message, $code);
} else {
    parent::__construct((string)$message, $code, $previous);
}
```

File: framework/helpers/BaseStringHelper.php [hunk 11]
```diff
@@ -436,7 +436,7 @@ public static function matchWildcard($pattern, $string, $options = [])
             $pattern .= 'i';
         }
 
-        return preg_match($pattern, $string) === 1;
+        return preg_match($pattern, isset($string) ? $string : '') === 1;
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php:439]: Typecast instead.

File: framework/helpers/BaseStringHelper.php [hunk 12]
```diff
@@ -450,6 +450,7 @@ public static function matchWildcard($pattern, $string, $options = [])
      */
     public static function mb_ucfirst($string, $encoding = 'UTF-8')
     {
+        $string = isset($string) ? $string : '';
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php:453]: Typecast instead.

File: framework/validators/FileValidator.php [hunk 13]
```diff
@@ -188,15 +188,17 @@ public function init()
             $this->tooSmall = Yii::t('yii', 'The file "{file}" is too small. Its size cannot be smaller than {formattedLimit}.');
         }
         if (!is_array($this->extensions)) {
-            $this->extensions = preg_split('/[\s,]+/', strtolower($this->extensions), -1, PREG_SPLIT_NO_EMPTY);
+            $fileExtension = isset($this->extensions) ? $this->extensions : "";
```
Comment (samdark) [inline framework/validators/FileValidator.php:191]: Typecast instead.

File: framework/validators/FileValidator.php [hunk 14]
```diff
@@ -188,15 +188,17 @@ public function init()
...
+            $this->extensions = preg_split('/[\s,]+/', strtolower($fileExtension), -1, PREG_SPLIT_NO_EMPTY);
         } else {
             $this->extensions = array_map('strtolower', $this->extensions);
         }
         if ($this->wrongMimeType === null) {
             $this->wrongMimeType = Yii::t('yii', 'Only files with these MIME types are allowed: {mimeTypes}.');
         }
         if (!is_array($this->mimeTypes)) {
-            $this->mimeTypes = preg_split('/[\s,]+/', strtolower($this->mimeTypes), -1, PREG_SPLIT_NO_EMPTY);
+            $mimeType = isset($this->mimeTypes) ? $this->mimeTypes : "";
```
Comment (samdark) [inline framework/validators/FileValidator.php:200]: Typecast instead.

File: framework/web/GroupUrlRule.php [hunk 15]
```diff
@@ -77,7 +77,7 @@ class GroupUrlRule extends CompositeUrlRule
      */
     public function init()
     {
-        $this->prefix = trim($this->prefix, '/');
+        $this->prefix = trim(isset($this->prefix) ? $this->prefix : '', '/');
```
Comment (samdark) [inline framework/web/GroupUrlRule.php:80]: Typecast instead.

File: framework/web/Request.php [hunk 16]
```diff
@@ -570,7 +570,7 @@ public function getBodyParams()
             }
 
             $rawContentType = $this->getContentType();
-            if (($pos = strpos($rawContentType, ';')) !== false) {
+            if (($pos = strpos(isset($rawContentType) ? $rawContentType : '', ';')) !== false) {
```
Comment (samdark) [inline framework/web/Request.php:573]: Typecast instead.

File: framework/web/UrlNormalizerRedirectException.php [hunk 17]
```diff
@@ -47,6 +47,7 @@ public function __construct($url, $statusCode = 302, $scheme = false, $message =
         $this->url = $url;
         $this->scheme = $scheme;
         $this->statusCode = $statusCode;
+        $message = isset($message) ? $message : '';
```
Comment (samdark) [inline framework/web/UrlNormalizerRedirectException.php:50]: Typecast instead.

File: tests/framework/filters/auth/BasicAuthTest.php [hunk 18]
```diff
@@ -78,7 +78,7 @@ public function testHttpBasicAuthCustom($token, $login)
         $filter = [
             'class' => HttpBasicAuth::className(),
             'auth' => function ($username, $password) {
-                if (preg_match('/\d$/', $username)) {
+                if (preg_match('/\d$/', isset($username) ? $username : '')) {
```
Comment (samdark) [inline tests/framework/filters/auth/BasicAuthTest.php:81]: Typecast instead.

Comment (longthanhtran) [reply tests/framework/filters/auth/BasicAuthTest.php]: thanks, and updated with typecast

File: tests/framework/base/SecurityTest.php [hunk 19]
```diff
@@ -1046,6 +1046,10 @@ public function dataProviderDeriveKey()
      */
     public function testHkdf($hash, $ikm, $salt, $info, $l, $prk, $okm)
     {
+        $hash = isset($hash) ? $hash : '';
+        $ikm  = isset($ikm) ? $ikm : '';
+        $salt = isset($salt) ? $salt : '';
+        $info = isset($info) ? $info : '';
         $dk = $this->security->hkdf($hash, hex2bin($ikm), hex2bin($salt), hex2bin($info), $l);
```
[the remark this answers is not in the document]
Comment (longthanhtran) [reply tests/framework/base/SecurityTest.php]: thanks, and updated with typecast

File: framework/web/UrlNormalizerRedirectException.php (hunk 17 above)
Comment (longthanhtran) [reply framework/web/UrlNormalizerRedirectException.php]: thanks, and updated with typecast

File: framework/web/Response.php [hunk 20]
```diff
@@ -897,7 +897,8 @@ public function redirect($url, $statusCode = 302, $checkAjax = true)
 
         if ($checkAjax) {
             if ($request->getIsAjax()) {
-                if (in_array($statusCode, [301, 302]) && preg_match('/Trident\/|MSIE[ ]/', $request->userAgent)) {
+                $userAgent = isset($request->userAgent) ? $request->userAgent : '';
```
[the remark this answers is not in the document]
Comment (longthanhtran) [reply framework/web/Response.php]: thanks, and updated with typecast

File: framework/web/Request.php [hunk 21]
```diff
@@ -796,7 +796,8 @@ public function setHostInfo($value)
     public function getHostName()
     {
         if ($this->_hostName === null) {
-            $this->_hostName = parse_url($this->getHostInfo(), PHP_URL_HOST);
+            $hostInfo = null !== $this->getHostInfo() ? $this->getHostInfo() : '';
```
[the remark this answers is not in the document]
Comment (longthanhtran) [reply framework/web/Request.php]: thanks, and updated with typecast

File: framework/validators/FilterValidator.php [hunk 22]
```diff
@@ -75,6 +75,7 @@ public function validateAttribute($model, $attribute)
     {
         $value = $model->$attribute;
         if (!$this->skipOnArray || !is_array($value)) {
+            $value = isset($value) ? $value : '';
```
[the remark this answers is not in the document]
Comment (longthanhtran) [reply framework/validators/FilterValidator.php]: thanks, and updated with typecast

File: framework/validators/FileValidator.php (hunk 14 above)
Comment (longthanhtran) [reply framework/validators/FileValidator.php]: thanks, and updated with typecast

File: framework/validators/FileValidator.php (hunk 13 above)
Comment (longthanhtran) [reply framework/validators/FileValidator.php]: thanks, and updated with typecast

File: framework/helpers/BaseStringHelper.php (hunk 12 above)
Comment (longthanhtran) [reply framework/helpers/BaseStringHelper.php]: thanks, and updated with typecast

File: framework/helpers/BaseStringHelper.php (hunk 11 above)
Comment (longthanhtran) [reply framework/helpers/BaseStringHelper.php]: thanks, and updated with typecast

**Review by samdark — APPROVED** [review]

**Pull request merged**